### PR TITLE
CI Integration

### PR DESCRIPTION
**Pull request by brynblack**
### Description

This PR integrates the new GitHub Actions CI/CD system into Polykey-CLI's development pipeline. It replaces GitLab's previous CI system that we were utilising (`.gitlab-ci.yml`). The new system is situated under the path `.github/workflows/ci.yml` and contains the GitHub Actions workflow that governs this entire repository.

It should implement the same feature set as the GitLab CI system that we were using, and be able to:
- Run automated checks
- Open PRs from staging to master
- Automatically fast-forward merge into master
- Automatically deploy (this will be changed in the future to use Orchestrator)
- Create pre-releases
- Create releases

### Issues Fixed

* Relates to [INF-32](https://linear.app/matrix-ai/issue/INF-32/ci-migration-from-gitlab-to-the-next-generation-ci)

### Tasks

- [x] 1. `feature*` branches:
  1. Performs linting.
  2. Does a dry run.
  3. Tests the build process.
  4. Generates & spawns child jobs that run all of the tests under `check-test`.
- [x] 2. `staging` branch:
  1. Runs lint, dry, and generate for all the tests. Same as the feature branches.
  2. Does a build and stores the artifacts in `./dist`.
  3. Creates a merge PR from `staging` into `master`.
  4. Does the initial build with Nix. This creates the application, docker and linux builds. They are saved as artifacts in `./builds/` and `./dist/`.
  5. It then does the testnet deployment. Deploying the testnet to the ECR. And waits for the deployment using the `wait-for-deploy.js` script.
  6. It the does integration testing for all of the platforms. For docker, linux, macos, nix and windows. It attempts a sanity check run of all the builds here. Notably the windows and macos runners will do the `pkg` step before testing. These pkged executables are included in the artifacts.
  7. Finally it will merge the PR into master.
- [x] 3. `prerelease` tag, will do all of basic `staging` branch jobs but also do a prerelease to NPM. Should only run in staging.
- [x] 4. `release` tag. This does all of the steps in the staging branch with the additional steps.
  1. Will upload the docker artifact to the ECR for the mainnet. And wait for the deployment with the `wait-for-deploy.js` script.
  2. Does a final sanity check of the docker image by running tests against it.
  3. Finally triggers a GH release and deploys the docker image to the ECR.

### Final checklist

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+
+  # Other
+  GH_PROJECT_PATH: ${{ github.repository }}
+  GIT_SUBMODULE_STRATEGY: recursive
+  npm_config_cache: "${{ github.workspace }}/tmp/npm"
+  npm_config_prefer_offline: "true"
+
+jobs:
+  check-lint:
+    name: "Check / Lint"
```
Comment (tegefaulkes) [inline .github/workflows/ci.yml:26]: Naming should be consistent with what it was before. So something like `check:lint` here. Unless ':' can't be used for some reason. Also keep it all lowercase.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+            --assignee "@me" \
+            --no-maintainer-edit || true
+
+          printf "Pipeline Attempt on $GITHUB_RUN_ID for $GITHUB_SHA\n\n$GITHUB_SERVER_URL/$GITHUB_REPOSITORY/actions/runs/$GITHUB_RUN_ID" \
+          | gh pr comment staging \
+            --body-file - \
+            --repo "$GH_PROJECT_PATH"
+
+  build-dist:
+    name: "Build / Dist"
```
Comment (aryanjassal) [inline .github/workflows/ci.yml]: Perhaps rename `Build / Dist` to something like `Build / Distribution` to make it more obvious of it's purpose.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+          echo 'Publishing library prerelease'
+
+          nix develop .#ci --command bash -c $'
+            npm publish --tag prerelease --access public
+          '
+
+          rm -f ./.npmrc
+
+  integration-builds:
+    name: "Integration / Builds"
```
Comment (aryanjassal) [inline .github/workflows/ci.yml:176]: Same here. Perhaps rename this to `Integration / Run Builds`?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+
+  integration-deployment:
+    name: "Integration / Deployment"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    concurrency:
+      group: integration-deployment
+      cancel-in-progress: true
```
Comment (aryanjassal) [inline .github/workflows/ci.yml:230]: Why is this needed here? Why do we want to cancel in-progress deployments if a new one is triggered?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+            ./scripts/wait-for-deploy.js testnet.polykey.com
+          '
+
+  integration-platforms:
+    name: "Integration / Platforms"
+    runs-on: ubuntu-latest
+
+    concurrency:
+      group: integration-platforms
+      cancel-in-progress: true
```
Comment (aryanjassal) [inline .github/workflows/ci.yml]: Same here. Why should in-progress testing on platforms be cancelled if a new one is triggered?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+  integration-prerelease:
+    name: "Integration / Pre-release"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    concurrency:
+      group: integration-prerelease
+      cancel-in-progress: true
```
Comment (aryanjassal) [inline .github/workflows/ci.yml]: Same here. If a new release is pushed up to staging, why should an old pre-release be cancelled?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+
+  integration-merge:
+    name: "Integration / Merge"
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/matrixai/github-runner
+
+    concurrency:
+      group: integration-merge
+      cancel-in-progress: true
```
Comment (aryanjassal) [inline .github/workflows/ci.yml:414]: Same here. If a new commit is pushed up to staging, why are we cancelling the merge?

**Review by aryanjassal — COMMENTED** [review]
Everything looks good except for minor changes. I still am not sure why we need to cancel in-progress jobs if a new commit triggers the CI.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (brynblack) [reply .github/workflows/ci.yml:230]: Concurrency groups were previously used in the `.gitlab-ci.yml` file in the form of resource groups, so this is being mirrored here using concurrency groups instead. This is intentional.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+      - name: Build dist
+        run: |
+          echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > ./.npmrc
+          echo 'Publishing library prerelease'
+
+          nix develop .#ci --command bash -c $'
+            npm publish --tag prerelease --access public
+          '
+
+          rm -f ./.npmrc
```
Comment (tegefaulkes) [inline .github/workflows/ci.yml]: The release of a `prerelease` tag should only happen on tagged commits that are tagged as a prerelease.

So If I did a `npm version prerelease --preid=alpha` to create the prerelease tag, the pushed that up it should run on that. Any other kind of commit should not run this job.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+    name: "Integration / Platforms"
+    runs-on: ubuntu-latest
+
+    concurrency:
+      group: integration-platforms
+      cancel-in-progress: true
+
+    needs: integration-builds
+
+    if: github.ref == 'refs/heads/staging'
```
Comment (tegefaulkes) [inline .github/workflows/ci.yml]: The `integration-platforms` job where we test all of the platforms needs to depend on the `testnet` deployment in `integration-deployment`. We need to wait for the testnet seednodes to update before we proceed with the integration testing.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+    concurrency:
+      group: integration-prerelease
+      cancel-in-progress: true
+
+    needs:
+      - build-prerelease
+      - integration-platforms
+      - integration-deployment
+
+    if: github.ref == 'refs/heads/staging'
```
Comment (tegefaulkes) [inline .github/workflows/ci.yml]: This also should only run on pre-release tagged commits.

File: .github/workflows/ci.yml
Comment (tegefaulkes) [inline .github/workflows/ci.yml:1]: I think we're completely missing the release tag jobs. How are the mainnet seed nodes deployed, an NPM release published and a GitHub release done?

**Review by tegefaulkes — COMMENTED** [review]
I'm focusing on the structure and making sure all the steps done in the gitlab CI were covered. Looks like you did a good job simplifying things here but I think you're missing the proper release stages when we push a release tag.

Also, you don't seem to be checking for version tags at all? They are used to trigger releases via NPM and github releases. So we need that logic here somewhere.

I went through the gitlab CI for some examples and I've written up a summary.

1. CI should never run on an un-protected branch.
2. `feature*` branches, these are protected branches. The do basic checks and tests
	1. checks lint.
	2. dry run.
	3. test-generate - spawns children jobs that run all of the tests under `check:test`.
3. `staging` branch, does all the basic tests but also merges.
	1. Runs lint, dry, and generate for all the tests. Same as the feature branch
	2. Does a build and keeps the artifacts in `./dist`
	3. creates a merge PR
	4. Does the initial build with nix. This creates the application, docker and linux builds. They are saved as artifacts in `./builds/` and `./dist/`
	5. It then does the testnet deployment. Deploying the testnet to the ECR. And waits for the deployment using the `wait-for-deploy.js` script.
	6. It the does integration testing for all of the platforms. For docker, linux, macos, nix and windows. It attempts a sanity check run of all the builds here. Notably the windows and macos runners will do the `pkg` step before testing. These `pkg`ed executables are included in the artefacts.
	7. Finally it will merge the PR into master.
4. prerelease tag, will do all of basic staging branch jobs but also do a prerelease to NPM. Should only run in staging.
5. release tag. This does all of the steps in the `staging` branch with the additional steps.
	1. Will upload the docker artifact to the ECR for the mainnet. And wait for the deployment with the `wait-for-deploy.js` script.
	2. Does a final sanity check of the docker image by running tests against it.
	3. Finally triggers a GH release and deploys the docker image to the ECR.

Also, how are protected branches handled? The CI should only run on the protected branches. That should be `feature*`, `staging` and `master`. This is important to prevent any un-trusted people from pushing whatever they want and executing random code in the CI.

**Comment by CMCDragonkai** [comment]
> Everything looks good except for minor changes. I still am not sure why we need to cancel in-progress jobs if a new commit triggers the CI.

This is necessary to avoid useless jobs. And having loads of jobs starting from each commit.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (brynblack) [reply .github/workflows/ci.yml:26]: It is consistent, I just decided to use `-` instead of `:` because its more in line with how YAML keys are named. And everything is lowercase still. The other part `name:`, that is just the display name on GitHub, it does not change anything, but looks more readable. Example:
image

File: .github/workflows/ci.yml (hunk 8 above)
Comment (brynblack) [reply .github/workflows/ci.yml]: Should now only target commits that begin with the tag `v`, so that would include stuff like `v0.5.4` or whatever.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (brynblack) [reply .github/workflows/ci.yml]: Should now depend on `integration-deployment`.

File: .github/workflows/ci.yml (hunk 10 above)
Comment (brynblack) [reply .github/workflows/ci.yml]: Done.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (brynblack) [reply .github/workflows/ci.yml]: Eh its probably fine as is.

**Comment by CMCDragonkai** [comment]
This needs a more expansive task list based on @tegefaulkes review @brynblack.

**Comment by CMCDragonkai** [comment]
Can't tick off everything until it's squashed and all features and behaviour replicated (unless the original behaviour was buggy).

File: .github/workflows/ci.yml
Comment (brynblack) [reply .github/workflows/ci.yml:1]: This should be done.

**Comment by tegefaulkes** [comment]
I don't have too much more to add. Except maybe take the numbered list from my review comment. Convert it to a check list and check off the criteria as a sanity check.

**Comment by brynblack** [comment]
The only thing missing now seems to be the `check-test` job that generates test files and runs them on `feature*` branches. Otherwise everything else seems to be implemented.

**Comment by brynblack** [comment]
Ready for merge. Going to merge to test the deployment system.
